### PR TITLE
Add user resource management to the task API

### PR DESCRIPTION
**Pull request by lyondhill**
step 1 of #11427

### CODE REVIEW

File: http/task_service.go [hunk 1]
```diff
@@ -554,6 +567,17 @@ func (h *TaskHandler) handleDeleteTask(w http.ResponseWriter, r *http.Request) {
 		EncodeError(ctx, err, w)
 		return
 	}
+	// clean up resource maps for deleted task
+	urms, _, err := h.UserResourceMappingService.FindUserResourceMappings(ctx, platform.UserResourceMappingFilter{
+		ResourceID:   req.TaskID,
+		ResourceType: platform.TasksResourceType,
+	})
+
+	if err == nil {
```
Comment (mark-rushakoff) [inline http/task_service.go]: We should do something when err is non-nil. Maybe just `log.Warn` it?

File: http/task_service_test.go [hunk 2]
```diff
@@ -760,3 +760,64 @@ func TestTaskHandler_NotFoundStatus(t *testing.T) {
...
+		},
+	}
+	h.handlePostTask(w, r)
+	r = httptest.NewRequest("DELETE", "http://any.url/api/v2/tasks/"+taskID.String(), nil)
+
+	h.ServeHTTP(w, r)
+
+	if created.UserID != deletedUser {
+		t.Fatalf("deleted user (%s) doesn't match created user (%s)", deletedUser, created.UserID)
+
```
Comment (mark-rushakoff) [inline http/task_service_test.go]: Remove blank line.

File: http/task_service_test.go [hunk 3]
```diff
@@ -760,3 +760,64 @@ func TestTaskHandler_NotFoundStatus(t *testing.T) {
...
+		OrganizationID: 1,
+	}
+
+	b, err := json.Marshal(task)
+	if err != nil {
+		t.Fatalf("failed to unmarshal task: %v", err)
+	}
+
+	r := httptest.NewRequest("POST", "http://any.url/v1", bytes.NewReader(b))
+	ctx := pcontext.SetAuthorizer(context.TODO(), &platform.Authorization{UserID: 2})
```
Comment (mark-rushakoff) [inline http/task_service_test.go]: `context.Background()` instead of `context.TODO` here.

File: http/task_service.go [hunk 4]
```diff
@@ -350,6 +350,19 @@ func (h *TaskHandler) handlePostTask(w http.ResponseWriter, r *http.Request) {
...
 	}
 
+	urm := &platform.UserResourceMapping{
+		UserID:       auth.GetUserID(),
+		UserType:     platform.Owner,
+		ResourceType: platform.TasksResourceType,
+		ResourceID:   req.Task.ID,
+	}
+	if err := h.UserResourceMappingService.CreateUserResourceMapping(ctx, urm); err != nil {
+		err = &platform.Error{
```
Comment (mark-rushakoff) [inline http/task_service.go]: Need to do something with this error. Offhand I don't know what pattern is used in the other parts of the codebase if this call fails.

Comment (lyondhill) [reply http/task_service.go]: Yah I will look around. I dont know if we delete the newly created task here or just return an error response?

File: http/task_service.go [hunk 5]
```diff
@@ -350,6 +350,29 @@ func (h *TaskHandler) handlePostTask(w http.ResponseWriter, r *http.Request) {
...
+		UserID:       auth.GetUserID(),
+		UserType:     platform.Owner,
+		ResourceType: platform.TasksResourceType,
+		ResourceID:   req.Task.ID,
+	}
+	if err := h.UserResourceMappingService.CreateUserResourceMapping(ctx, urm); err != nil {
+		// clean up the task if we fail to map the user and resource
+		// TODO(lh): Multi step creates could benefit from a service wide transactional request
+		if derr := h.TaskService.DeleteTask(ctx, req.Task.ID); derr != nil {
+			err = fmt.Errorf("%s: failed to clean up task", err.Error())
```
Comment (mark-rushakoff) [inline http/task_service.go]: Let's make sure `derr` is reflected here too. 

I think there's a way you can chain platform.Error together, if you make a platform.Error out of `err` and then associate `derr` somehow.

File: http/task_service.go [hunk 6]
```diff
@@ -554,6 +577,19 @@ func (h *TaskHandler) handleDeleteTask(w http.ResponseWriter, r *http.Request) {
...
+	urms, _, err := h.UserResourceMappingService.FindUserResourceMappings(ctx, platform.UserResourceMappingFilter{
+		ResourceID:   req.TaskID,
+		ResourceType: platform.TasksResourceType,
+	})
+
+	if err != nil {
+		h.logger.Warn("failed to pull user resource mapping", zap.Error(err))
+	} else {
+		for _, m := range urms {
+			h.UserResourceMappingService.DeleteUserResourceMapping(ctx, m.ResourceID, m.UserID)
```
Comment (mark-rushakoff) [inline http/task_service.go]: DeleteUserResourceMapping returns an error too, and should be logged if non-nil.

**Comment by bednar** [comment]
Hi @lyondhill 

When I trying create the Task via HTTP then the response is unsuccessful with an error:

```bash
-> POST 
-> http://127.0.0.1:9999/api/v2/tasks
-> {
->	"flux": "option task = {\n    name: \"it task1549451552382-IT\",\n    every: 1h\n}\n\nfrom(bucket:\"my-bucket\") |> range(start: 0) |> last()",
->	"links": {},
->	"name": "it task1549451552382-IT",
->	"orgID": "035c9e6a90ff9000",
->	"status": "active"
-> }

<- RESPONSE
<- {"code":"internal error","message":"failed to add user permissions","error":{"code":"","message":"unsupported resource type tasks"}}
```

The problem is same as in #11491.

Can you look at it?

**Comment by mark-rushakoff** [comment]
@bednar the task creation error you mentioned is most likely fixed by https://github.com/influxdata/influxdb/pull/11701, which is still WIP.

**Comment by bednar** [comment]
@mark-rushakoff I can confirm that it works

**Pull request merged**